Fixed issue with KernelPCA.inverse_transform mean

lrjball opened the pull request:
Currently `KernelPCA.inverse_transform` returns a data set with zero mean, even if the original data did not have zero mean.

I believe that this PR fixes that issue, so that the mean of the inverse-transformed data set is the same as the mean of the original data set. 

I've also added a test for this update.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #16654

lrjball committed:
Fixed issue with KernelPCA.inverse_transform mean

Added the mean to the inverse_transform to fix the issue, and have added tests as well.

lrjball committed:
fixed linting issue

jnothman commented on sklearn/decomposition/tests/test_kernel_pca.py:
```diff
@@ -282,3 +283,15 @@ def test_kernel_conditioning():
+    kp = KernelPCA(n_components=2, kernel=kernel, fit_inverse_transform=True)
+    X_trans = kp.fit_transform(X)
+    X_inv = kp.inverse_transform(X_trans)
+    assert np.isclose(X.mean(axis=0), X_inv.mean(axis=0)).all()
```
Should we be able to assert re the values, not just the means?

lrjball replied:
Good point, I've updated the test now.

jnothman reviewed (commented):
Thanks

Please add an entry to the change log at `doc/whats_new/v0.23.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

Please also note versionchanged in the docstring of inverse_transform

lrjball committed:
Added to changelog, removed mean from tests and fixed behaviour

- Added entry in whats new v0.23
- Updated tests to check for closeness of full data set, not just closeness of the mean.
- Updated the fix for this bug. Realised that it wasn't an issue with the mean not being added on, but instead that self.alpha was not being taken into account in the inverse transform.

lrjball commented:
I realized that the fix to the bug wasn't actually due to the mean not being added, but was due to the alpha not being handled properly in the inverse_transform. So I have updated the fix and have also added an entry in the change log.

lrjball committed:
Updated changelog

jnothman commented on doc/whats_new/v0.23.rst:
```diff
@@ -138,6 +138,10 @@ Changelog
   :func:`decomposition.non_negative_factorization` now preserves float32 dtype.
   :pr:`16280` by :user:`Jeremie du Boisberranger <jeremiedbb>`.
 
+- |Fix| :class:`decomposition.KernelPCA` method ``inverse_transform`` now
```
Maybe could be "fixed .... in the case that data was not centred" would be more helpful to users

lrjball replied:
So actually after doing some digging, it was still returning the wrong thing even for centered data. I just only noticed the bug in non-centered data because the mean of the inverse-transformed data was zero when the original data set was not centered.

For example, in 0.22.0 the following still does not work:
```
import numpy as np
from sklearn.datasets import make_blobs
from sklearn.decomposition import KernelPCA

X, _ = make_blobs(n_samples=100, centers=[[1, 1, 1, 1]], random_state=0)
X = X - X.mean(axis=0)
kp = KernelPCA(n_components=2, fit_inverse_transform=True)
X_trans = kp.fit_transform(X)
X_inv = kp.inverse_transform(X_trans)

assert np.isclose(X, X_inv).all()
```

So this PR fixes the `inverse_transform` function for all X.

jnothman reviewed (commented):
Thanks for the update

glemaitre commented on sklearn/decomposition/tests/test_kernel_pca.py:
```diff
@@ -282,3 +283,15 @@ def test_kernel_conditioning():
+    kp = KernelPCA(n_components=2, kernel=kernel, fit_inverse_transform=True)
+    X_trans = kp.fit_transform(X)
+    X_inv = kp.inverse_transform(X_trans)
+    assert np.isclose(X, X_inv).all()
```
you can use `assert_allclose`: `from sklearn.utils._testing import assert_allclose`

```suggestion
    assert_allclose(X, X_inv)
```

glemaitre reviewed (commented):
It seems correct. Just a small change in the test to make an assert on the numpy array.

lrjball committed:
Update sklearn/decomposition/tests/test_kernel_pca.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jnothman reviewed (approved):
Thank you, @lrjball!

jnothman commented:
@glemaitre, does this have your approval?

glemaitre commented:
@lrjball Thanks for the fix

kstoneriv3 commented:
For me, it seems that this PR broke the code. Could you check #18902?

kstoneriv3 commented:
> I realized that the fix to the bug wasn't actually due to the mean not being added, but was due to the alpha not being handled properly in the inverse_transform. ...

Though it is concluded here that the alpha is handled correctly, it is indeed handled correctly in the original implementation. The problem was that the mean was not added to the data when the kernel is linear. When the linear kernel is used, the information of the mean is completely lost by the centering of the kernel, while when non-linear kernel is used, the information of the mean is partially lost.

kstoneriv3 commented:
@jnothman @glemaitre
As discussed in #18902, I believe this PR introduced a bug, so I created the above PR (#19732) to fix it. I would appreciate it if you could take the time to review it.
